libmesabox/src/lsb/mount/mod.rs

```diff
@@ -0,0 +1,803 @@
+
+/// This is used to read Filesystem Description File
+/// e.g. /etc/fstab and /etc/mtab
+struct FSDescFile {
+    entries: Vec<MntEnt>,
+}
+
+impl FSDescFile {
+    fn new(path: &str) -> MountResult<Self> {
+        // all of these files should exist and can be read, but just in case
+        let file = fs::File::open(path).or(Err(MountError::OpenFileError(String::from(path))))?;
+        let mut entries = Vec::new();
+        for line in BufReader::new(file).lines() {
+            let line = line.or(Err(MountError::FSDescFileFormatError(String::from(path))))?;
+            match line.chars().next() {
+                None | Some('#') => {
+                    continue;
+                }
+                Some(_) => {}
+            }
+            let a: Vec<&str> = line.split_whitespace().collect();
+            // There should be 6 columns in FileSystem Description Files
+            if a.len() != 6 {
+                return Err(MountError::FSDescFileFormatError(String::from(path)));
+            }
+            // We only need the first 4 columns
+            let mnt = MntEnt {
+                mnt_fsname: OsString::from(a[0]),
+                mnt_dir: OsString::from(a[1]),
+                mnt_type: OsString::from(a[2]),
+                mnt_opts: OsString::from(a[3]),
+            };
+            entries.push(mnt)
+        }
+        Ok(Self { entries: entries })
+    }
+
+    fn get_output(&self, fs_type: &OsString) -> String {
+        let mut ret = String::new();
+        for item in &self.entries {
+            if *fs_type != *"" {
+                if *fs_type != item.mnt_type {
+                    continue;
+                }
+            }
+            ret.push_str(
+                format!(
+                    "{} on {} type {} ({})\n",
+                    item.mnt_fsname.to_string_lossy(),
+                    item.mnt_dir.to_string_lossy(),
+                    item.mnt_type.to_string_lossy(),
+                    item.mnt_opts.to_string_lossy(),
+                ).as_str(),
+            );
+        }
+        ret
+    }
+}
+
+/// Define some special requirements
+struct Property {
+    fake: bool,
+    use_uuid: bool,
+    use_label: bool,
+    // TODO user mountable devices
+}
+
+impl Default for Property {
+    fn default() -> Self {
+        Self {
+            fake: false,
+            use_uuid: false,
+            use_label: false,
+        }
+    }
+}
+
+/// OsString has limited methods, implement them
+trait OsStringExtend {
+    fn starts_with(&self, pat: &str) -> bool;
+    fn contains(&self, pat: &str) -> bool;
+}
+
+impl OsStringExtend for OsString {
+    fn starts_with(&self, pat: &str) -> bool {
+        self.to_string_lossy().starts_with(pat)
+    }
+
+    fn contains(&self, pat: &str) -> bool {
+        self.to_string_lossy().contains(pat)
+    }
+}
+
+impl MountOptions {
+    fn from_matches(matches: &clap::ArgMatches) -> MountResult<Self> {
+        let mut mount_list: Vec<Box<Mountable + Send>> = Vec::new();
```

Comment:
I would recommend trying to find another way to store the data, as this setup requires dynamic dispatch.  If you can't, I suppose it's fine for this utility as this part doesn't need to be super fast.

Reply:
Now I defined an enum to do the dispatching.